src/metrics/include/metrics/v2/event.hpp

```diff
@@ -0,0 +1,282 @@
+        _stub = poplar::PoplarEventCollector::NewStub(channel);
+        return true;
+    }
+
+    void setStub() {
+        static bool stub_created = createStub();
+        return;
+    }
+};
+
+
+/**
+ * Manages the stream of poplar EventMetrics.
+ */
+class StreamInterfaceImpl {
+public:
+    StreamInterfaceImpl()
+        : _options{}, _response{}, _context{}, _stream{_stub->StreamEvents(&_context, &_response)} {
+        _options.set_no_compression().set_buffer_hint();
+    }
+
+    void write(const poplar::EventMetrics& event) {
+        auto success = _stream->Write(event, _options);
+
+        if (!success) {
+            BOOST_THROW_EXCEPTION(PoplarRequestError("Failed to write to stream."));
+        }
+    }
+
+    ~StreamInterfaceImpl() {
+        if (!_stream) {
+            BOOST_LOG_TRIVIAL(error) << "Tried to close grpc stream, but none existed.";
+            return;
+        }
+        if (!_stream->WritesDone()) {
+            BOOST_LOG_TRIVIAL(warning) << "Closing grpc stream, but not all writes completed.";
+        }
+        auto status = _stream->Finish();
+        if (!status.ok()) {
+            BOOST_LOG_TRIVIAL(error)
+                << "Problem closing grpc stream: " << _context.debug_error_string();
+        }
+    }
+
+private:
+    CollectorStubInterface _stub;
+    grpc::WriteOptions _options;
+    poplar::PoplarResponse _response;
+    grpc::ClientContext _context;
+    std::unique_ptr<grpc::ClientWriterInterface<poplar::EventMetrics>> _stream;
+};
+
+/**
+ * Very basic mock for tests.
+ */
+class MockStreamInterface {
+public:
+    MockStreamInterface() {}
+
+    void write(const poplar::EventMetrics& event) {
+        events.push_back(event);
+    }
+
+    // We make this static so we can access it even several private objects deep.
+    static std::vector<poplar::EventMetrics> events;
+};
+
+
+/**
+ * Manages the gRPC-side collector for each operation.
+ * Exists for construction / destruction resource management only.
+ */
+class Collector {
+public:
+    Collector(const Collector&) = delete;
+
+    explicit Collector(const std::string& name, const std::string& path_prefix)
+        : _name{name}, _id{} {
+        _id.set_name(_name);
+
+        grpc::ClientContext context;
+        poplar::PoplarResponse response;
+        poplar::CreateOptions options = createOptions(_name, path_prefix);
+        auto status = _stub->CreateCollector(&context, options, &response);
+
+        if (!status.ok()) {
+            std::ostringstream os;
+            os << "Collector status not okay: " << status.error_message();
+            BOOST_THROW_EXCEPTION(PoplarRequestError(os.str()));
+        }
+    }
+
+    ~Collector() {
+        grpc::ClientContext context;
+        poplar::PoplarResponse response;
+        auto status = _stub->CloseCollector(&context, _id, &response);
+        if (!status.ok()) {
+            BOOST_LOG_TRIVIAL(error) << "Couldn't close collector: " << status.error_message();
```

Comment:
here and elsewhere, see if it's feasible to give names or more context in error messages - it may be helpful to know *which* collector fails to start/stop when debugging by looking at the state of the ftdc files dir etc.